rfc-124-deploying-banners.md

```diff
@@ -0,0 +1,81 @@
+
+![The current mechanism for deploying banners on GOV.UK](https://docs.google.com/drawings/d/e/2PACX-1vS-vz5bPpGAOhATHYH6hqY5BSR_T-JPSJkuGeneuZKe6Ae8Z4vShcPw-5Im9_OESK1zapFUZX4gNoX4/pub?w=1504&amp;h=909)
+
+The current mechanism for deploying banners on GOV.UK – [view source](https://docs.google.com/drawings/d/1XgSskX5Ufb6BU_PiaIvAHsj7bTARMMgmWlkfOf3f_uA/edit)
+
+## Problem
+
+The recent events – such as making the cookie banner Privacy and Electronic Communications Regulations (PECR) compliant or showing important information related to Brexit and more recently COVID-19 – have made us more aware about the importance of being able to deploy banners quickly and with confidence without overloading our infrastructure. As described in the previous section, the process for deploying banners on GOV.UK is not consistent and has a certain degree of complexity depending on the banner type.
+
+Moving away from `static`/`slimmer` towards a fully component-based architecture makes the process or rolling out a change across all applications more difficult.
+
+## Proposal
+
+Use Edge Side Include (ESI) to deploy banners on GOV.UK.
+
+Edge Side Include (ESI) is a [web standard](https://www.w3.org/TR/esi-lang) originally proposed by Akamai and Oracle, among other companies. It allows an Edge Server (like [Fastly’s caches](https://www.fastly.com/blog/using-esi-part-1-simple-edge-side-include)) to "mix and match" content from multiple URLs.
+
+In practice, the emergency banner on GOV.UK, for example, will require the following changes:
+
+1. make the emergency banner available as a fragment (e.g. on `/emergency-banner`); this should be hosted on a frontend server, potentially served by a new application called `banners`, similarly with the other public-facing application
+1. add an ESI include tag (pointing to the fragment) in the application layout
+```
+  <esi:include src="/emergency-banner" />
+```
+1. add an ESI remove tag and ESI comment to provide a fallback on environments where ESI is not available
+```
+  <esi:remove>
+    <%= render "/components/emergency_banner" %>
+  </esi:remove>
+  <!--esi
+  <esi:include src="/emergency-banner" />
+  -->
+```
+1. enable ESI in Fastly via Varnish Configuration Language (VCL)
+1. optional: add extra logic in VCL. Fastly gives us the ability to create custom responses at the edge – called "synthetics" in Varnish terminology – which can be used to include personalised, dynamic information that will be served from the edge instead of origin
+
+### Pros
+- performance gains (a higher cache-hit ratio, quicker page loads, less traffic spikes)
```

Comment:
Is the "less traffic spikes"/"higher cache-hit ratio" because the banner can be cachebusted in isolation, as per point 3 in this list? And is the "quicker page loads" because of the reduced JS?

If so, I think it's worth splitting these out into separate points ("improved server-side performance", "improved client-side performance")

Reply:
Updated this point and split it up as suggested. For the client-side, I'm referring to the fact that the cached page assets can still be used by the browser when the banner is enabled/disabled as they won't be re-fingerprinted.